File: Data Structures & Algorithms/valid-parenthesis-string/submission-0.js
class Solution {
    /**
     * @param {string} s
     * @return {boolean}
     */
    checkValidString(s) {
        let leftStk = [], starStk = [];
        for (let i = 0; i < s.length; i++) {
            let c = s[i];
            switch(c) {
                case "(":
                    leftStk.push(i);
                    break;
                case "*":
                    starStk.push(i);
                    break;
                default:
                    if (leftStk.length != 0) {
                        leftStk.pop();
                        continue;
                    }
                    if (starStk.length != 0) {
                        starStk.pop();
                        continue;
                    }
                    return false;
            }
        }
        while (leftStk.length != 0) {
            if (starStk.length == 0 || leftStk.pop() > starStk.pop()) return false;
        }
        return true;
    }
}

File: Data Structures & Algorithms/validate-parentheses/submission-0.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for c in s:
            match c:
                case ")":
                    if not stack or stack[-1] != "(":
                        return False
                    stack.pop()
                case "]":
                    if not stack or stack[-1] != "[":
                        return False
                    stack.pop()
                case "}":
                    if not stack or stack[-1] != "{":
                        return False
                    stack.pop()
                case _:
                    stack.append(c)
        
        return len(stack) == 0